cpp/include/cuml/random_projection/rproj_c.h

```diff
@@ -54,15 +57,26 @@ struct rand_mat {
   ~rand_mat() { this->reset(); }
 
   // For dense matrices
-  math_t *dense_data;
+  device_buffer<math_t> *dense_data;
 
   // For sparse CSC matrices
-  int *indices;
-  int *indptr;
-  math_t *sparse_data;
+  device_buffer<int> *indices;
```

Comment:
We can reduce duplication here by re-using the CSR object in `src_prims/sparse/CSR.h`. This will also encapsulate the ownership of the device buffiers in one place.

Reply:
Do you really want the `CSR` to replace the `rand_mat`, or simply use a `CSR` for the sparse part of the `rand_mat` class? Because, aren't `CSR` objects meant to store only sparse matrices. I know it's technically possible to also store a dense matrix in it, but it is not very satisfying solution. I made a change to have the device_buffers on the stack. I think there's really no risk of leaks here, unless the `rand_mat` object is on the heap and not properly released (same problem with `CSR`).